README.md
# nuWindow Promises

[![Build Status](https://img.shields.io/travis/electblake/nuwindow-promises.svg?branch=master)](http://travis-ci.org/docpad/docpad "Check this project's build status on TravisCI")

create popup windows with promises support (great for social sharing)

### Features

* Uses es6 promises
* pass options
* named windows


## Install
coming soon
## Usage
coming soon
## Help
coming soon


CHANGELOG.md
## 0.1.0 (2015-04-16)


#### Features

* **gulp:** migrated to gulp quickltime because gulp is dope. ([58ddc535](https://github.com/electblake/nuwindow-promises/commit/58ddc53596dcf18d0d1354e7b1130c86863df6ba))


### 0.0.8 (2014-12-15)


#### Bug Fixes

* **package:** renamed package to match repo to match install folder et al. ([8a8f3b99](https://github.com/electblake/nuwindow-promises/commit/8a8f3b996e1e28831f7d1459d86f74f3658b00b2))


### 0.0.7 (2014-12-15)


#### Features

* **position:** added automatic window centering ([3d24a506](https://github.com/electblake/nuwindow-promises/commit/3d24a506d61bbfb775e142fcf6b34468547a5244))


### 0.0.6 (2014-12-15)


#### Features

* **position:** windows now open centered on screen ([e7a5a31a](https://github.com/electblake/nuwindow-promises/commit/e7a5a31ac660a660c516f33b3e96e8ab90754fcb))


### 0.0.3 (2014-12-12)


#### Bug Fixes

* **devops:** added release-it config, and updated git dotfile ignores ([cb94618f](https://github.com/electblake/nuwindow-promises/commit/cb94618f6cfb9574b4baf4172a9596f330f4202b))


### 0.0.2 (2014-12-12)


#### Bug Fixes

* **devops:** updated package definitions ([5bae296c](https://github.com/electblake/nuwindow-promises/commit/5bae296c06b4dc41adba1fc72fafdb9d7624c7b1))


#### Features

* **promises:** added promises support and some module loading boilerplate ([4ef71946](https://github.com/electblake/nuwindow-promises/commit/4ef71946e705533c92701385457627d1f08254dc))



test/test.js
var expect = chai.expect;

describe('nuWindow', function() {
	var thisInst = null;
	var options = { };
	var name = 'nuwindow-test1';
	var url = 'http://github.com/electblake/nuwindow-promises';

	it('should be factory function', function() {
		expect(nuWindow).to.be.a('function');
	});

	var thisInst = nuWindow(name, url, options);
	// console.log('nuWindow', nuWindow);
	it('should return Promise', function() {
		expect(thisInst).to.be.an.instanceof(Promise);
	});

	it('should have window property', function() {
		expect(thisInst.window).to.be.an('object');
	});

	it('should close', function() {
		thisInst.window.close();
	});
});

describe('nuWindow Promise', function() {
	var options = {};
	var name = 'nuwindow-test2';
	var url = 'http://github.com/electblake/nuwindow-promises';

	it('should promise then', function() {
		var thisInst = nuWindow(name, url, options);
		// console.log('thisInst', thisInst);
		thisInst.then(function(result) {
			expect(result).to.exist;
		});
		thisInst.window.close();
	});

	it('should promise catch', function() {
		var thisInst = nuWindow(name, url, options);
		// console.log('thisInst', thisInst);
		thisInst.catch(function(result) {
			expect(result).to.exist;
		});
		thisInst.window.close();
	});
});

Gruntfile.js
// Generated on 2014-11-11 using generator-angular 0.9.8
'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// use this if you want to recursively match all subfolders:
// 'test/spec/**/*.js'

module.exports = function (grunt) {

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Configurable paths for the application
  var appConfig = {
    src: require('./bower.json').appPath || 'src',
    js: require('./bower.json').appPath || 'js',
    dist: 'dist'
  };

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    yeoman: appConfig,

    // Make sure code styles are up to par and there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.src %>/scripts/{,*/}*.js'
        ]
      },
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },

    watch: {
      build: {
        files: [
          '<%= yeoman.src %>/ng/**/*.js'
        ],
        tasks: [
          'angular:build'
        ]
      }
    },

    bump: {  
      options: {
        files: ["bower.json", "package.json"],
        commitFiles: ["-a"],
        commit: true,
        tagName: '%VERSION%',
        push: true,
        pushTo: 'origin'
      }
    },

    // Empties folders to start fresh
    clean: {
      build: {
        dot: true,
        src: [
          '<%= yeoman.js %>/{,*/}*'
        ]
      },
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/{,*/}*',
            '!<%= yeoman.dist %>/.git*'
          ]
        }]
      }
    },

    concat: {
      options: {
        separator: ';'
      },
      build: {
        src: [
          '<%= yeoman.src %>/**/*.js'
        ],
        dest: '<%= yeoman.js %>/nuwindow-promise.js',
      }
    },

    uglify: {
      build: {
        options: {
          sourceMap: true,
          sourceMapName: '<%= yeoman.dist %>/nuwindow-promise.min.js.map'
        },
        files: {
          '<%= yeoman.dist %>/nuwindow-promise.min.js': [
            '<%= yeoman.js %>/nuwindow-promise.js'
          ]
        }
      }
    },

    // Test settings
    karma: {
      unit: {
        configFile: 'test/karma.conf.js',
        singleRun: true
      }
    }
  });

  grunt.registerTask('build', [
    'clean:build',
    'concat:build',
    'uglify:build'
  ]);

};


karma.conf.js
// Karma configuration
// Generated on Thu Apr 16 2015 00:52:36 GMT-0400 (EDT)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha'],


    // list of files / patterns to load in the browser
    files: [
      'bower_components/es5-shim/es5-shim.min.js',
      'bower_components/es6-shim/es6-shim.min.js',
      'bower_components/lodash/dist/lodash.min.js',
      'bower_components/chai/chai.js',
      'src/*.js',
      'test/*.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['mocha'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


js/nuwindow-promise.js
// if the module has no dependencies, the above pattern can be simplified to
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define([], factory);
    } else if (typeof exports === 'object') {
        // Node. Does not work with strict CommonJS, but
        // only CommonJS-like environments that support module.exports,
        // like Node.
        module.exports = factory();
    } else {
        // Browser globals (root is window)
        root.nuWindow = factory();
  }
}(this, function () {

    // Just return a value to define the module export.
    // This example returns an object, but the module
    // can return a function as the exported value.
    var _ = window._;
    var nuWindow = function(name, url, options) {

        var _defaults = {
          width: 520,
          height: 350
        };

        options = _.defaults(_defaults, options);
        
        var _position = {
            left: (screen.width/2) - (options.width / 2),
            top: (screen.height/2) - (options.height / 2)
        };
        options = _.extend(options, _position);

        var specs = [];
        if (options) {
            _.forIn(options, function(element, index) {
                specs.push(index+'='+element);
            });
        }

        var optString = specs.join(', ');
        
        console.log('nuWindow optString', optString);
        var thisWindow = window.open(url, name, optString);

        var promise = new Promise(function(resolve, reject) {
            try {
              var interval = setInterval(function() {
                  if (thisWindow == null || thisWindow.closed) {
                      clearInterval(interval);
                      resolve(thisWindow);
                  }
              }, 1000);
            } catch (err) {
                reject(err);
            }
        });
        
        return promise;
    };

    return nuWindow;
}));

/**
 * 
 */

// var opts = {
//   width: 520,
//   height: 350
// };

// var tweetShareUrl = new URI('http://twitter.com/share');

// tweetShareUrl.addQuery('url', this.shareURL);
// tweetShareUrl.addQuery('text', this.shareText);
// tweetShareUrl.addQuery('via', 'getmusicl');

// var tweetURL = tweetShareUrl.toString();

// nuWindow(this._id + 'shareTwitter', tweetURL, opts)
//   .then(function() {
//     $log.debug('-- nuWindow finished');
//   })

src/nuWindow.js
// if the module has no dependencies, the above pattern can be simplified to
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define([], factory);
    } else if (typeof exports === 'object') {
        // Node. Does not work with strict CommonJS, but
        // only CommonJS-like environments that support module.exports,
        // like Node.
        module.exports = factory();
    } else {
        // Browser globals (root is window)
        root.nuWindow = factory();
  }
}(this, function () {

    // Just return a value to define the module export.
    // This example returns an object, but the module
    // can return a function as the exported value.
    // var _ = window._;

    if (!_) {
        throw new Error('[nuWindow] Lodash is required.');
    }

    function nuWindow(name, url, options) {

        var _defaults = {
            width: 520,
            height: 350
        };

        this.options = _.defaults(_defaults, options);
        
        var _position = {
            left: (screen.width/2) - (this.options.width / 2),
            top: (screen.height/2) - (this.options.height / 2)
        };

        this.options = _.extend(this.options, _position);

        var specs = [];
        if (this.options) {
            _.forIn(this.options, function(element, index) {
                specs.push(index+'='+element);
            });
        }

        var optString = specs.join(', ');
        
        console.info('[nuWindow] specs', optString);

        var thisWindow = window.open(url, name, optString);

        var promise = new Promise(function(resolve, reject) {
            try {
              var interval = setInterval(function() {
                  if (thisWindow == null || thisWindow.closed) {
                      clearInterval(interval);
                      resolve(thisWindow);
                  }
              }, 1000);
            } catch (err) {
                reject(err);
            }
        });

        promise.window = thisWindow;
        
        return promise;
    };

    return nuWindow;
}));

/**
 * 
 */

// var opts = {
//   width: 520,
//   height: 350
// };

// var tweetShareUrl = new URI('http://twitter.com/share');

// tweetShareUrl.addQuery('url', this.shareURL);
// tweetShareUrl.addQuery('text', this.shareText);
// tweetShareUrl.addQuery('via', 'getmusicl');

// var tweetURL = tweetShareUrl.toString();

// nuWindow(this._id + 'shareTwitter', tweetURL, opts)
//   .then(function() {
//     $log.debug('-- nuWindow finished');
//   })